src/aoai-api-simulator/src/aoai_api_simulator/telemetry.py

```diff
@@ -0,0 +1,148 @@
+            name="aoai-api-simulator.latency.full",
+            description="Full latency of handling the request (including simulated latency)",
+            unit="seconds",
+        ),
+        # dimensions: deployment, token_type
+        histogram_tokens_used=meter.create_histogram(
+            name="aoai-api-simulator.tokens.used",
+            description="Number of tokens used per request",
+            unit="tokens",
+        ),
+        # dimensions: deployment, token_type
+        histogram_tokens_requested=meter.create_histogram(
+            name="aoai-api-simulator.tokens.requested",
+            description="Number of tokens across all requests (success or not)",
+            unit="tokens",
+        ),
+        # dimensions: deployment
+        histogram_tokens_rate_limit=meter.create_histogram(
+            name="aoai-api-simulator.tokens.rate-limit",
+            description="Number of tokens that were counted for rate-limiting",
+            unit="tokens",
+        ),
+        # dimensions: deployment, reason
+        histogram_rate_limit=meter.create_histogram(
+            name="aoai-api-simulator.limits",
+            description="Number of requests that were rate-limited",
+            unit="requests",
+        ),
+    )
+
+
+simulator_metrics = _get_simulator_metrics()
+
+
+def setup_telemetry() -> bool:
+    using_azure_monitor: bool
+
+    if application_insights_connection_string:
+        logger.info("🚀 Configuring Azure Monitor telemetry")
+
+        # Options: https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/monitor/azure-monitor-opentelemetry#usage
+        configure_azure_monitor(connection_string=application_insights_connection_string)
+        using_azure_monitor = True
+    else:
+        using_azure_monitor = False
+        logger.info("Azure Monitor telemetry not configured (set APPLICATIONINSIGHTS_CONNECTION_STRING)")
+
+    if opentelemetry_exporter_otlp_endpoint:
+        logger.info("🚀 Configuring OTLP telemetry")
+
+        # setup the instrumentors
+        resource = Resource(attributes={"service.name": os.getenv("OTEL_SERVICE_NAME", "aoai-api-simulator")})
+
+        # tracing
+        if not using_azure_monitor:
+            trace.set_tracer_provider(TracerProvider(resource=resource))
+
+        span_processor = BatchSpanProcessor(OTLPSpanExporter())
+        trace.get_tracer_provider().add_span_processor(span_processor)
+
+        # metrics
+        metric_reader = PeriodicExportingMetricReader(OTLPMetricExporter())
+
+        if not using_azure_monitor:
+            meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
+            metrics.set_meter_provider(meter_provider)
+        else:
+            meter_provider = metrics.get_meter_provider()
+            # meter_provider.add_metric_reader() is not implemented in python sdk yet.
+            # adding it manually
+            meter_provider._all_metric_readers.add(metric_reader)
```

Comment:
I'm trying to understand these two lines.

The first is adding the metric reader. Is there a reason that we need to add the reader to the current meter rather than creating a `MeterProvider` as above?

I'm less clear on the intent/purpose of the second line - can you explain?

Reply:
Because if the `configure_azure_monitor(connection_string=application_insights_connection_string)` already sets up the meter provider and the meter provider API lacks a method to add a new metric reader to an existing instance of a meter provider.

Reply:
basically if the azure monitor library sets up the meter provider first, our only option is to go under the covers and add a metric reader this way.

Reply:
I don't recall having anything like this previously with the `configure_azure_monitor` call. What has change that means we need to add this now?

Reply:
Because we have to construct the OTLP exporter pipeline now for both the scenario of...

1. Azure monitor has already setup the providers
2. Azure monitor not having setup the providers (no env variable for connection string)

The if else logic addresses those 2 scenarios.

Azure monitor has it's own pipeline and OTLP has it's own, but both need to share the same providers.

Reply:
So in case 1 (Azure Monitor has already set up the providers), this previously didn't need these calls. What has changed in this PR that means that we now need to access these private methods where we didn't before?

Reply:
Because we didn't have a need to setup OTLP exporters before. We do now.

Reply:
tl;dr

logic before...
```
if azure_monitor_connection_string has been provided:
   setup_azure_monitor()
```   

logic now...
```
if azure_monitor_connection_string has been provided:
   setup_azure_monitor()
   pipelines_already_constructed = true
   
if OTLP_endpoint has been provided:
  if pipelines_already_constructed:
     don't construct a new provider, use existing
  else:
    construct a new provider
 ```

Reply:
Ah - I think I was missing a configuration in my mind as I was processing this.

Scenarios:

1. Send metrics to App Insights only, not using OLTP pipeline. This is the case prior to this PR and this would use the `setup_azure_monitor` call and wouldn't need the added code that uses the private members
2. Send metrics to OLTP and don't use the App Insights package. This would create the new provider + OTLP pipeline
3. Send metrics to App Insights and export via OTLP. 
 
The last scenario is the one that I hadn't previously had in mind, and I believe that it is the one where the private member calls are required?

Reply:
> The last scenario is the one that I hadn't previously had in mind, and I believe that it is the one where the private member calls are required?

Yes.

There is also the scenario where none of the two are enabled in which case we don't do anything.